daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -1004,24 +1008,47 @@ private[lf] object SBuiltin {
           else
             machine.returnValue = contract
         case None =>
-          throw SpeedyHungry(
-            SResultNeedContract(
-              coid,
-              templateId,
-              onLedger.committers,
-              cbMissing = _ => machine.tryHandleSubmitMustFail(),
-              cbPresent = { case V.ContractInst(actualTmplId, V.VersionedValue(_, arg), _) =>
-                // Note that we cannot throw in this continuation -- instead
-                // set the control appropriately which will crash the machine
-                // correctly later.
-                machine.ctrl =
-                  if (actualTmplId != templateId)
-                    SEDamlException(DamlEWronglyTypedContract(coid, templateId, actualTmplId))
-                  else
-                    SEImportValue(typ, arg)
-              },
-            )
-          )
+          onLedger.cachedContracts.get(coid) match {
```

Comment:
We need to check the that coid has the proper type.

Reply:
Added in 5d3a7eb2274f0cebbc499e768391ae94113502c0 and added a test in cee52077c3a510f6e4e713bfd5af0db284363a87 which fails without this change.